GTBitsOfGood/southface
# hc/92-recent-standards

## 92-recent-standards

Issue Number(s): #92 

Adds recentStandards property to User which is updated when a user views standards. These are then displayed next to standard card components on plan builder pages with granular control for the number of standards to see recency of. The request itself is an API call that also has that control.

### Checklist

**Frontend**
- [x] recent standards component created
- [x] recent standards component takes in a prop that specifies how many standards to track + show
- [x] track default 3 standards at the moment
       - the prop should allow you to track more than three standards but default it to 3
       - consider using a queue both for the frontend state and backend database (add to front + remove from back) (see other notes)
- [x] the appropriate standard card component is rendered in the recent section and is scrollable (overflow: scroll)
- [x] onClick on standard opens up the modal for that specific standard (the `CardModal` component) (see other notes)
- [x] onClick on "see all standards" routes back to `/library` page

**Backend**
- [x] array of type card is added to user schema (call it something like "recentStandardsQueue")
- [x] onClick of each standard in the library page, it will update the queue both in the frontend and backend
- [ ] once max number of standards to be displayed is reached (based on the prop), remove from end of array

### How to Test

- Log in
- Click on a few cards in the library. Have fun with it.
- Go to Project Plan Builder (The only place that loads RecentStandardsView in this PR)
- Go to saved plans.
- See if the ones you picked are loaded in the right order. First to last. 
- Click See All Standards. It should take you back to the library page.

### Notes

SWR updated to 2.0 to work with post requests.

StandardCard on open updates the users recent standards, but doesn’t really take advantage of useSWRMutation. Caching is unnecessary. Regardless, useSWRMutation is still used because that’s what the issue said to do.

getIds expects a POST request, not a GET request. This is because the number of ids can be quite large.

The recentStandardsComponent is created (RecentStandardsView) but requires recentStandards to be passed to it as is done in saved-project-plans.jsx right now. This can be easily changed to have a useUser hook in the component itself, I just wasn’t sure on how useSWR does caching and figured it was better to be safe.

There is no cap on the number of standards that get stored right now. I don’t see it as a problem because the amount of data that could potentially be added is negligible in comparison to the side of the project, but let me know if that’s a must have. It could require more database queries.

updateRecentStandards in the server-side User actions file requires potentially two queries if the standard is not already a recent standard. I tried merging them but couldn’t figure it out after 2 hours of testing. If you can get it to work, amazing, but it’s wonky because you’re not working on a collection, but a property of an object in a collection.

The recentStandardsView does NOT re-render when a standard is opened on the same page. It does update, however. Reloading shows the change. Good, bad? Idk. That’s a design decision to be made, I guess. To update it, look into revalidating useSWR.
